core-plugins/src/main/java/co/cask/hydrator/plugin/common/KinesisUtil.java

```diff
@@ -0,0 +1,132 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.common;
+
+import com.amazonaws.services.kinesis.AmazonKinesisClient;
+import com.amazonaws.services.kinesis.model.CreateStreamRequest;
+import com.amazonaws.services.kinesis.model.DescribeStreamRequest;
+import com.amazonaws.services.kinesis.model.ResourceNotFoundException;
+import com.amazonaws.services.kinesis.model.StreamStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Utility methods to help create and manage Kinesis Streams
+ */
+public final class KinesisUtil {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KinesisUtil.class);
+  private static final long TIMEOUT = 3 * 60 * 1000;
+  private static final long SLEEP_INTERVAL = 1000 * 10;
+
+  private KinesisUtil() {
+  }
+
+  /**
+   * Creates an Amazon Kinesis stream if it does not exist and waits for it to become available
+   *
+   * @param kinesisClient The {@link AmazonKinesisClient} with Amazon Kinesis read and write privileges
+   * @param streamName The Amazon Kinesis stream name to create
+   * @param shardCount The shard count to create the stream with
+   * @throws IllegalStateException Invalid Amazon Kinesis stream state
+   * @throws IllegalStateException Stream does not become active before the timeout
+   */
+  public static void createAndWaitForStream(AmazonKinesisClient kinesisClient, String streamName, int shardCount) {
+    if (streamExists(kinesisClient, streamName)) {
+      StreamStatus streamStatus = getStreamState(kinesisClient, streamName);
+      switch (streamStatus) {
+        case DELETING:
+          waitForStreamState(kinesisClient, streamName, null);
+          createStream(streamName, shardCount, kinesisClient);
+          break;
+        case ACTIVE:
+          LOG.info("Stream {} already exists", streamName);
+          return;
+        case CREATING:
+          LOG.info("Stream {} is being created", streamName);
+          break;
+        case UPDATING:
+          LOG.info("Stream {} is being updated", streamName);
+          return;
+        default:
+          throw new IllegalStateException(String.format("Illegal stream state: %s", streamStatus));
+      }
+    } else {
+      createStream(streamName, shardCount, kinesisClient);
+    }
+    waitForStreamState(kinesisClient, streamName, StreamStatus.ACTIVE);
+  }
+
+  private static void waitForStreamState(AmazonKinesisClient kinesisClient, String streamName,
+                                         @Nullable StreamStatus expectedStatus) {
+    long waitTime = System.currentTimeMillis() + TIMEOUT;
+    while (System.currentTimeMillis() < waitTime) {
+      try {
+        TimeUnit.MILLISECONDS.sleep(SLEEP_INTERVAL);
```

Comment:
With this logic, it means it will always sleep for at least 10 seconds even before checking the status? Also, sleeping 10 seconds seems awe fully long